### ISSUE TITLE
apiserver panic'ed due to concurrent map read/write

### ISSUE THREAD

**Issue by ppruthi** [kubernetes/kubernetes#64484]
**Is this a BUG REPORT or FEATURE REQUEST?**:
Bug Report

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind feature

**What happened**:
Kube API server panics due to `fatal error: concurrent map read and map write`

**What you expected to happen**:
API server to handle concurrent writes

**How to reproduce it (as minimally and precisely as possible)**:
- Kubernetes 1.9.7 on AWS
- Cluster runs kube-dns, kube-dns-autoscalar, cluster-autoscalar, kube2iam, rescheduler, etc.
- Running a few workloads with ImagePullBackOff failure hitting the pods

**Anything else we need to know?**:
Kube API server log:
```
fatal error: concurrent map read and map write
goroutine 239937 [running]:
runtime.throw(0x3b0d043, 0x21)
        /usr/local/go/src/runtime/panic.go:605 +0x95 fp=0xc432ac9900 sp=0xc432ac98e0 pc=0x42c285
runtime.mapaccess2_faststr(0x334ece0, 0xc445632f90, 0x3aba126, 0x9, 0xc426e1abe8, 0xc63d817619ed0401)
        /usr/local/go/src/runtime/hashmap_fast.go:324 +0x47a fp=0xc432ac9958 sp=0xc432ac9900 pc=0x40c56a
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.nestedFieldNoCopy(0xc445632f60, 0xc432ac9a68, 0x2, 0x2, 0x310b980, 0xc426ad9340, 0xc432ac9a01
        /workspace/anago-v1.9.7-beta.0.111+dd5e1a2978fd0b/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.NestedString(0xc445632f60, 0xc432ac9a68, 0x2, 0x2, 0xc4288a3440, 0x3b, 0xc445632f01)
        /workspace/anago-v1.9.7-beta.0.111+dd5e1a2978fd0b/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.getNestedString(0xc445632f60, 0xc432ac9a68, 0x2, 0x2, 0xc4288a3440, 0x3b)
        /workspace/anago-v1.9.7-beta.0.111+dd5e1a2978fd0b/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured.(*Unstructured).GetNamespace(0xc4311da830, 0xc4288a3440, 0x3b)
...
..
.
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage.(*timeBudget).periodicallyRefresh(0xc4209d5f00, 0xc420059740)
        /workspace/anago-v1.9.7-beta.0.111+dd5e1a2978fd0b/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/time_bud
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage.newTimeBudget
        /workspace/anago-v1.9.7-beta.0.111+dd5e1a2978fd0b/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/time_bud
goroutine 378 [select, 25 minutes]:
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage.(*Cacher).dispatchEvents(0xc4205eb450)
        /workspace/anago-v1.9.7-beta.0.111+dd5e1a2978fd0b/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.g
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage.NewCacherFromConfig
        /workspace/anago-v1.9.7-beta.0.111+dd5e1a2978fd0b/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.g
goroutine 379 [select, 4 minutes]:
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*Reflector).watchHandler(0xc420b2cbe0, 0x8e44b60, 0xc443ced560, 0xc42cd0bbc8, 0xc443ced380, 0xc420059740, 0x0, 0x
        /workspace/anago-v1.9.7-beta.0.111+dd5e1a2978fd0b/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/reflecto
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch(0xc420b2cbe0, 0xc420059740, 0x0, 0x0)
        /workspace/anago-v1.9.7-beta.0.111+dd5e1a2978fd0b/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache/reflecto
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage.(*Cacher).startCaching(0xc4205eb450, 0xc420059740)
        /workspace/anago-v1.9.7-beta.0.111+dd5e1a2978fd0b/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.g
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage.NewCacherFromConfig.func1.1()
        /workspace/anago-v1.9.7-beta.0.111+dd5e1a2978fd0b/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher.g
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc42cd41fa8)
kube-apiserver.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
kube-apiserver.service: Unit entered failed state.
kube-apiserver.service: Failed with result 'exit-code'.
kube-apiserver.service: Service hold-off time over, scheduling restart.
Stopped Kubernetes API Server.
```

**Environment**:
- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.7", GitCommit:"dd5e1a2978fd0b97d9b78e1564398aeea7e7fe92", GitTreeState:"clean", BuildDate:"2018-04-19T00:05:56Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.7", GitCommit:"dd5e1a2978fd0b97d9b78e1564398aeea7e7fe92", GitTreeState:"clean", BuildDate:"2018-04-18T23:58:35Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

- Cloud provider or hardware configuration:
AWS

- OS (e.g. from /etc/os-release):
Fedora 27 (Cloud Edition)

- Kernel (e.g. `uname -a`):
4.16.7-200.fc27.x86_64

- Install tools:

- Others:

**Comment by wenjiaswe** [comment]
I think there is a race condition in k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage.(*Cacher).dispatchEvents, and this PR [#57366](https://github.com/kubernetes/kubernetes/pull/57366) should have fixed this, but the fix is not in v1.9. @mbohlool , would you please help to cherry pick  PR #57366 to 1.9? Thanks!

**Comment by wenjiaswe** [comment]
Wait for PR #65002 to merge.
